src/psyclone/tests/domain/lfric/dofkern_test.py

```diff
@@ -142,3 +151,119 @@ def test_dof_kernel_invalid_field_vector():
+                        arg_type(gh_field, gh_real, gh_read, w1)   &
+        """,
+        1)
+    ast = fpapi.parse(code, ignore_comments=False)
+    name = "testkern_dofs_type"
+    # Load the metadata into an empty kernel
+    md = LFRicKernMetadata(ast, name=name)
+    kern = LFRicKern()
+    kern.load_meta(ktype=md)
+    # Assert that the identifier is set
+    assert kern.is_dofkern
+
+
+def test_multiple_write_args():
+    '''
+    Check that LFRicKernMetadata raises no errors when user-defined dof kernel
+    defines more than one metadata arg of type 'gh_field' with an access type
+    of 'gh_write'.
+
+    '''
+    # Substitute field for multiple writes
+    code = CODE.replace(
+        """
+                type(arg_type), dimension(2) :: meta_args =        &
+                    (/ arg_type(gh_field, gh_real, gh_write, w1),  &
+                        arg_type(gh_field, gh_real, gh_read, w2)   &
+        """,
+        """
+                type(arg_type), dimension(3) :: meta_args =        &
+                    (/ arg_type(gh_field, gh_real, gh_write, w1),  &
+                        arg_type(gh_field, gh_real, gh_write, w1), &
+                        arg_type(gh_field, gh_real, gh_read, w1)   &
+        """,
+        1)
+    ast = fpapi.parse(code, ignore_comments=False)
+    name = "testkern_dofs_type"
+    # Load the metadata into an empty kernel
+    md = LFRicKernMetadata(ast, name=name)
+    # Proxy assertion that validating dof kern passes
+    assert md.is_dofkern
+
+
+def test_upper_bound_undf():
+    '''
+    Checks that the correct upper bound is generated for a dof-kernel when
+    distributed memory is set to 'False'. This should be set to the unique
+    number of dofs for the single function space in the subroutine, denoted
+    by the 'undf' variable.
+
+    '''
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "1.14_single_invoke_dofs.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=False).create(invoke_info)
+    code = str(psy.gen)
+
+    expected = ("      loop0_start = 1\n"
+                "      loop0_stop = undf_w1")
+
+    assert expected in code
+
+
+def test_upper_bound_dofowned():
+    '''
+    Checks that the correct upper bound is generated for a dof-kernel when
+    distributed memory is set to 'True'. This should be set to the last dof
+    owned by the vector space and be accessed by the 'get_last_dof_owned'
+    subroutine.
+
+    '''
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "1.14_single_invoke_dofs.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    code = str(psy.gen)
+
+    expected = ("      loop0_start = 1\n"
+                "      loop0_stop = f1_proxy%vspace%get_last_dof_owned()")
+
+    assert expected in code
+
```

Comment:
I think we also need to test the behaviour when 'annexed dofs' is switched on too. Again, there's an `annexed` fixture for this. See https://psyclone-dev.readthedocs.io/en/latest/APIs.html#dof-iterators for more details.